### PR TITLE
Spring cleaning: Remove Kore.AST.MetaOrObject

### PR DESCRIPTION
**Pull request by ttuegel**
###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance: `stylish-haskell`

---

### CODE REVIEW

File: kore/src/Kore/Builtin/Set.hs [hunk 1]
```diff
@@ -603,32 +600,30 @@ unifyEquals
...
+                    setUnifier <- unifyEqualsChildren emptySetPat set2
                     -- error when subunification problem returns partial result
                     -- More details at 'errorIfIncompletelyUnified'.
                     errorIfIncompletelyUnified emptySetPat set2 setUnifier
                     -- Return the concrete set, but capture any predicates and
                     -- substitutions from unifying the element
                     -- and framing variable.
                     let result = pure dv1 <* elemUnifier <* setUnifier
-                    return (result, SimplificationProof)
+                    return (result)
```
Comment (marick) [inline kore/src/Kore/Builtin/Set.hs]: Nit: unneeded paren

File: kore/src/Kore/Builtin/Set.hs [hunk 2]
```diff
@@ -686,17 +677,17 @@ unifyEquals
...
-        -> unifier (expanded, proof)
+        -> TermLike variable  -- ^ key
+        -> unifier (Pattern variable)
     unifyEqualsElement builtin1 element' key2 =
         case Set.toList set1 of
             [fromConcreteStepPattern -> key1] ->
                 do
-                    (key, _) <- unifyEqualsChildren key1 key2
+                    key <- unifyEqualsChildren key1 key2
                     let result =
```
Comment (marick) [inline kore/src/Kore/Builtin/Set.hs]: Unnecessary binding. (These will all get caught and removed gradually, I assume, which is the way It Should Be.)

File: kore/src/Kore/Step/Axiom/EvaluationStrategy.hs [hunk 3]
```diff
@@ -261,7 +253,7 @@ applyFirstSimplifierThatWorks
     axiomIdToSimplifier
     patt
   = do
-    (applicationResult, _proof) <-
+    applicationResult <-
```
Comment (marick) [inline kore/src/Kore/Step/Axiom/EvaluationStrategy.hs]: Unneeded binding

File: kore/src/Kore/Step/Function/Evaluator.hs [hunk 4]
```diff
@@ -322,22 +311,20 @@ evaluateSortInjection tools ap
         in
             assert (toSort' == fromSort) $
                 ( resultApp
```
Comment (marick) [inline kore/src/Kore/Step/Function/Evaluator.hs]: unneeded paren

File: kore/src/Kore/Step/Function/Evaluator.hs [hunk 5]
```diff
@@ -322,22 +311,20 @@ evaluateSortInjection tools ap
...
-    updateSortInjectionSource head1 fromSort1 =
-        \children ->
-            ( Application
-                { applicationSymbolOrAlias =
-                    head1 { symbolOrAliasParams = [fromSort1, toSort1] }
-                , applicationChildren = children
-                }
-            )
+    updateSortInjectionSource head1 fromSort1 children =
+        Application
```
Comment (marick) [inline kore/src/Kore/Step/Function/Evaluator.hs]: I do wish internal structure like this were hidden behind a function interface.

File: kore/src/Kore/Step/Merging/Pattern.hs [hunk 6]
```diff
@@ -71,21 +69,18 @@ mergeWithPredicate
         , substitution = pattSubstitution
         }
   = do
-    (   Conditional
-            { predicate = mergedCondition
-            , substitution = mergedSubstitution
-            }
-        , _proof ) <-
-            mergePredicatesAndSubstitutions
-                tools
```
Comment (marick) [inline kore/src/Kore/Step/Merging/Pattern.hs]: I don't suppose you could rename MetadataTools while you're at it?

File: kore/src/Kore/Step/Simplification/AndTerms.hs [hunk 7]
```diff
@@ -739,34 +711,31 @@ transformerLiftOld
...
-    case first of
-        Bottom_ _ -> return (first, SimplificationProof)
-        Top_ _ -> return (second, SimplificationProof)
-        _ -> case second of
-            Bottom_ _ -> return (second, SimplificationProof)
-            Top_ _ -> return (first, SimplificationProof)
-            _ -> empty
+    -> Maybe (TermLike variable)
+boolAnd first second
+  | isBottom first  = return first
```
Comment (marick) [inline kore/src/Kore/Step/Simplification/AndTerms.hs]: Idle curiousity: Would this be idiomatic, vs. `Just first` and `Nothing`?

File: kore/src/Kore/Step/Axiom/Matcher.hs [hunk 8]
```diff
@@ -142,16 +139,16 @@ unificationWithAppMatchOnTop
...
-    -> BuiltinAndAxiomSimplifierMap Object
+    -> BuiltinAndAxiomSimplifierMap
     -- ^ Map from axiom IDs to axiom evaluators
     -> TermLike variable
     -> TermLike variable
     -> unifier
-        ( OrPredicate Object variable
-        , UnificationProof Object variable
+        ( OrPredicate variable
+
```
Comment (virgil-serbanuta) [inline kore/src/Kore/Step/Axiom/Matcher.hs]: Empty line not needed.

File: kore/src/Kore/Step/Axiom/Matcher.hs [hunk 9]
```diff
@@ -588,20 +581,20 @@ unifyJoin
...
     -- ^ Evaluates functions.
-    -> BuiltinAndAxiomSimplifierMap Object
+    -> BuiltinAndAxiomSimplifierMap
     -- ^ Map from axiom IDs to axiom evaluators
     -> [(TermLike variable, TermLike variable)]
     -> unifier
-        ( OrPredicate Object variable
-        , UnificationProof Object variable
+        ( OrPredicate variable
+
```
Comment (virgil-serbanuta) [inline kore/src/Kore/Step/Axiom/Matcher.hs]: Empty line not needed

File: kore/src/Kore/Step/Rule.hs [hunk 10]
```diff
@@ -110,23 +110,23 @@ rulePattern left right =
...
 
-deriving instance Eq variable => Eq (RewriteRule level variable)
-deriving instance Ord variable => Ord (RewriteRule level variable)
-deriving instance Show variable => Show (RewriteRule level variable)
+deriving instance Eq variable => Eq (RewriteRule variable)
+deriving instance Ord variable => Ord (RewriteRule variable)
+deriving instance Show variable => Show (RewriteRule variable)
 
-instance (Unparse variable, Ord variable) => Unparse (RewriteRule level variable) where
+instance (Unparse variable, Ord variable) => Unparse (RewriteRule variable) where
```
Comment (virgil-serbanuta) [inline kore/src/Kore/Step/Rule.hs]: Line length (was worse before, though)

File: kore/src/Kore/Unification/UnifierImpl.hs [hunk 11]
```diff
@@ -230,13 +189,13 @@ normalizeSubstitutionDuplication
...
-    -> BuiltinAndAxiomSimplifierMap Object
+    -> PredicateSimplifier
+    -> TermLikeSimplifier
+    -> BuiltinAndAxiomSimplifierMap
     -> Substitution variable
     -> unifier
-        ( Predicate Object variable
-        , UnificationProof Object variable
+        ( Predicate variable
+
```
Comment (virgil-serbanuta) [inline kore/src/Kore/Unification/UnifierImpl.hs]: Empty line not needed.

**Pull request merged**